File: server/camcops_server/tasks/bdi.py
#!/usr/bin/env python
# camcops_server/tasks/bdi.py

"""
===============================================================================

    Copyright (C) 2012-2018 Rudolf Cardinal (rudolf@pobox.com).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <http://www.gnu.org/licenses/>.

===============================================================================
"""

from typing import Any, Dict, List, Tuple, Type

from cardinal_pythonlib.stringfunc import strseq
import cardinal_pythonlib.rnc_web as ws
from sqlalchemy.ext.declarative import DeclarativeMeta
from sqlalchemy.sql.schema import Column
from sqlalchemy.sql.sqltypes import Integer, String

from camcops_server.cc_modules.cc_constants import (
    CssClass,
    DATA_COLLECTION_ONLY_DIV,
)
from camcops_server.cc_modules.cc_ctvinfo import CTV_INCOMPLETE, CtvInfo
from camcops_server.cc_modules.cc_db import add_multiple_columns
from camcops_server.cc_modules.cc_html import answer, bold, td, tr, tr_qa
from camcops_server.cc_modules.cc_request import CamcopsRequest
from camcops_server.cc_modules.cc_summaryelement import SummaryElement
from camcops_server.cc_modules.cc_task import Task, TaskHasPatientMixin
from camcops_server.cc_modules.cc_trackerhelpers import TrackerInfo


# =============================================================================
# Constants
# =============================================================================

BDI_I_QUESTION_TOPICS = {
    # from Beck 1988, https://doi.org/10.1016/0272-7358(88)90050-5
    1: "mood",  # a
    2: "pessimism",  # b
    3: "sense of failure",  # c
    4: "lack of satisfaction",  # d
    5: "guilt feelings",  # e
    6: "sense of punishment",  # f
    7: "self-dislike",  # g
    8: "self-accusation",  # h
    9: "suicidal wishes",  # i
    10: "crying",  # j
    11: "irritability",  # k
    12: "social withdrawal",  # l
    13: "indecisiveness",  # m
    14: "distortion of body image",  # n
    15: "work inhibition",  # o
    16: "sleep disturbance",  # p
    17: "fatigability",  # q
    18: "loss of appetite",  # r
    19: "weight loss",  # s
    20: "somatic preoccupation",  # t
    21: "loss of libido",  # u
}
BDI_IA_QUESTION_TOPICS = {
    # from [Beck1996b]
    1: "sadness",
    2: "pessimism",
    3: "sense of failure",
    4: "self-dissatisfaction",
    5: "guilt",
    6: "punishment",
    7: "self-dislike",
    8: "self-accusations",
    9: "suicidal ideas",
    10: "crying",
    11: "irritability",
    12: "social withdrawal",
    13: "indecisiveness",
    14: "body image change",
    15: "work difficulty",
    16: "insomnia",
    17: "fatigability",
    18: "loss of appetite",
    19: "weight loss",
    20: "somatic preoccupation",
    21: "loss of libido",
}
BDI_II_QUESTION_TOPICS = {
    # from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5889520/;
    # also https://www.ncbi.nlm.nih.gov/pubmed/10100838;
    # also [Beck1996b]
    # matches BDI-II paper version
    1: "sadness",
    2: "pessimism",
    3: "past failure",
    4: "loss of pleasure",
    5: "guilty feelings",
    6: "punishment feelings",
    7: "self-dislike",
    8: "self-criticalness",
    9: "suicidal thoughts or wishes",
    10: "crying",
    11: "agitation",
    12: "loss of interest",
    13: "indecisiveness",
    14: "worthlessness",
    15: "loss of energy",
    16: "changes in sleeping pattern",  # decrease or increase
    17: "irritability",
    18: "changes in appetite",  # decrease or increase
    19: "concentration difficulty",
    20: "tiredness or fatigue",
    21: "loss of interest in sex",
}
SCALE_BDI_I = "BDI-I"  # must match client
SCALE_BDI_IA = "BDI-IA"  # must match client
SCALE_BDI_II = "BDI-II"  # must match client
TOPICS_BY_SCALE = {
    SCALE_BDI_I: BDI_I_QUESTION_TOPICS,
    SCALE_BDI_IA: BDI_IA_QUESTION_TOPICS,
    SCALE_BDI_II: BDI_II_QUESTION_TOPICS,
}

NQUESTIONS = 21
TASK_SCORED_FIELDS = strseq("q", 1, NQUESTIONS)
MAX_SCORE = NQUESTIONS * 3
SUICIDALITY_QNUM = 9  # Q9 in all versions of the BDI (I, IA, II)
SUICIDALITY_FN = "q9"  # fieldname
CUSTOM_SOMATIC_KHANDAKER_BDI_II_QNUMS = [4, 15, 16, 18, 19, 20, 21]
CUSTOM_SOMATIC_KHANDAKER_BDI_II_FIELDS = Task.fieldnames_from_list(
    "q", CUSTOM_SOMATIC_KHANDAKER_BDI_II_QNUMS)


# =============================================================================
# BDI (crippled)
# =============================================================================

class BdiMetaclass(DeclarativeMeta):
    # noinspection PyInitNewSignature
    def __init__(cls: Type['Bdi'],
                 name: str,
                 bases: Tuple[Type, ...],
                 classdict: Dict[str, Any]) -> None:
        add_multiple_columns(
            cls, "q", 1, NQUESTIONS,
            minimum=0, maximum=3,
            comment_fmt="Q{n} [{s}] (0-3, higher worse)",
            comment_strings=[
                "BDI-I: {i}; BDI-IA: {ia}; BDI-II: {ii}".format(
                    i=BDI_I_QUESTION_TOPICS[q],
                    ia=BDI_IA_QUESTION_TOPICS[q],
                    ii=BDI_II_QUESTION_TOPICS[q],
                )
                for q in range(1, NQUESTIONS + 1)
            ]
        )
        super().__init__(name, bases, classdict)


class Bdi(TaskHasPatientMixin, Task,
          metaclass=BdiMetaclass):
    __tablename__ = "bdi"
    shortname = "BDI"
    longname = "Beck Depression Inventory (data collection only)"
    provides_trackers = True

    bdi_scale = Column(
        "bdi_scale", String(length=10),  # was Text
        comment="Which BDI scale (BDI-I, BDI-IA, BDI-II)?"
    )

    def is_complete(self) -> bool:
        return (
            self.field_contents_valid() and
            self.bdi_scale is not None and
            self.are_all_fields_complete(TASK_SCORED_FIELDS)
        )

    def get_trackers(self, req: CamcopsRequest) -> List[TrackerInfo]:
        return [TrackerInfo(
            value=self.total_score(),
            plot_label="BDI total score (rating depressive symptoms)",
            axis_label="Score for Q1-21 (out of {})".format(MAX_SCORE),
            axis_min=-0.5,
            axis_max=MAX_SCORE + 0.5
        )]

    def get_clinical_text(self, req: CamcopsRequest) -> List[CtvInfo]:
        if not self.is_complete():
            return CTV_INCOMPLETE
        return [CtvInfo(
            content="{} total score {}/{}".format(
                ws.webify(self.bdi_scale), self.total_score(), MAX_SCORE)
        )]

    def get_summaries(self, req: CamcopsRequest) -> List[SummaryElement]:
        return self.standard_task_summary_fields() + [
            SummaryElement(name="total",
                           coltype=Integer(),
                           value=self.total_score(),
                           comment="Total score (/{})".format(MAX_SCORE)),
        ]

    def total_score(self) -> int:
        return self.sum_fields(TASK_SCORED_FIELDS)

    def is_bdi_ii(self) -> bool:
        return self.bdi_scale == SCALE_BDI_II

    def get_task_html(self, req: CamcopsRequest) -> str:
        score = self.total_score()

        # Suicidal thoughts:
        suicidality_score = getattr(self, SUICIDALITY_FN)
        if suicidality_score is None:
            suicidality_text = bold("? (not completed)")
            suicidality_css_class = CssClass.INCOMPLETE
        elif suicidality_score == 0:
            suicidality_text = str(suicidality_score)
            suicidality_css_class = ""
        else:
            suicidality_text = bold(str(suicidality_score))
            suicidality_css_class = CssClass.WARNING

        # Custom somatic score for Khandaker Insight study:
        somatic_css_class = ""
        if self.is_bdi_ii():
            somatic_values = self.get_values(
                CUSTOM_SOMATIC_KHANDAKER_BDI_II_FIELDS)
            somatic_missing = False
            somatic_score = 0
            for v in somatic_values:
                if v is None:
                    somatic_missing = True
                    somatic_css_class = CssClass.INCOMPLETE
                    break
                else:
                    somatic_score += int(v)
            somatic_text = ("incomplete" if somatic_missing
                            else str(somatic_score))
        else:
            somatic_text = "N/A"  # not the BDI-II

        # Question rows:
        q_a = ""
        qdict = TOPICS_BY_SCALE.get(self.bdi_scale)
        topic = "?"
        for q in range(1, NQUESTIONS + 1):
            if qdict:
                topic = qdict.get(q, "??")
            q_a += tr_qa(
                "{question} {qnum} ({topic})".format(
                    question=req.wappstring("question"),
                    qnum=q,
                    topic=topic,
                ),
                getattr(self, "q" + str(q))
            )

        # HTML:
        h = """
            <div class="{CssClass.SUMMARY}">
                <table class="{CssClass.SUMMARY}">
                    {tr_is_complete}
                    {tr_total_score}
                    <tr>
                        <td>
                            Suicidal thoughts/wishes score
                            (Q{suicidality_qnum}) <sup>[1]</sup>
                        </td>
                        {td_suicidality}
                    </tr>
                    {tr_somatic_score}
                </table>
            </div>
            <div class="{CssClass.EXPLANATION}">
                All questions are scored from 0–3
                (0 free of symptoms, 3 most symptomatic).
            </div>
            <table class="{CssClass.TASKDETAIL}">
                <tr>
                    <th width="70%">Question</th>
                    <th width="30%">Answer</th>
                </tr>
                {tr_which_scale}
                {q_a}
            </table>
            <div class="{CssClass.FOOTNOTES}">
                [1] Suicidal thoughts are asked about in Q{suicidality_qnum} 
                    for all of: BDI-I (1961), BDI-IA (1978), and BDI-II (1996).
                    
                [2] Insight study: 
                    <a href="https://doi.org/10.1186/ISRCTN16942542">doi:10.1186/ISRCTN16942542</a>
                
                [3] See the <a href="http://camcops.org/documentation/tasks/bdi.html">CamCOPS 
                    BDI help</a> for full references and bibliography for the 
                    citations that follow.
                
                    <b>The BDI rates “right now” [Beck1988]. 
                    The BDI-IA rates the past week [Beck1988]. 
                    The BDI-II rates the past two weeks [Beck1996b].</b>
                    
                    1961 BDI(-I) question topics from [Beck1988].
                    1978 BDI-IA question topics from [Beck1996b].
                    1996 BDI-II question topics from [Steer1999], [Gary2018].
                </ul>

            </div>
            {data_collection_only_div}
        """.format(  # noqa
            CssClass=CssClass,
            tr_is_complete=self.get_is_complete_tr(req),
            tr_total_score=tr(
                req.wappstring("total_score"),
                answer(score) + " / {}".format(MAX_SCORE)
            ),
            suicidality_qnum=SUICIDALITY_QNUM,
            td_suicidality=td(suicidality_text, td_class=suicidality_css_class),  # noqa
            tr_somatic_score=tr(
                td(
                    "Custom somatic score for Insight study <sup>[2]</sup> "
                    "(sum of scores for questions {}, for BDI-II only)".format(
                        ", ".join("Q" + str(qnum) for qnum in
                                  CUSTOM_SOMATIC_KHANDAKER_BDI_II_QNUMS))
                ),
                td(somatic_text, td_class=somatic_css_class),
                literal=True
            ),
            tr_which_scale=tr_qa(
                req.wappstring("bdi_which_scale") + " <sup>[3]</sup>",
                ws.webify(self.bdi_scale)
            ),
            q_a=q_a,
            data_collection_only_div=DATA_COLLECTION_ONLY_DIV
        )
        return h

File: tablet_qt/tasks/bdi.cpp
/*
    Copyright (C) 2012, University of Cambridge, Department of Psychiatry.
    Created by Rudolf Cardinal (rnc1001@cam.ac.uk).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <https://www.gnu.org/licenses/>.
*/

#include "bdi.h"

#include "common/appstrings.h"
#include "common/textconst.h"
#include "lib/stringfunc.h"
#include "maths/mathfunc.h"
#include "questionnairelib/questionnaire.h"
#include "questionnairelib/questionwithonefield.h"
#include "questionnairelib/qumcq.h"
#include "questionnairelib/qumcqgrid.h"
#include "questionnairelib/qutext.h"
#include "tasklib/taskfactory.h"
#include "tasklib/taskregistrar.h"
using mathfunc::noneNull;
using mathfunc::sumInt;
using mathfunc::totalScorePhrase;
using stringfunc::strseq;

const int FIRST_Q = 1;
const int N_QUESTIONS = 21;
const int MAX_QUESTION_SCORE = N_QUESTIONS * 3;
const QString QPREFIX("q");

const int SUICIDALITY_QNUM = 9;  // Q9 in all versions of the BDI (I, IA, II)
const QString SUICIDALITY_FN = "q9";  // fieldname

const QVector<int> CUSTOM_SOMATIC_KHANDAKER_BDI_II_QNUMS{
    4, 15, 16, 18, 19, 20, 21};

const QString Bdi::BDI_TABLENAME("bdi");
const QString FN_BDI_SCALE("bdi_scale");

const QString SCALE_BDI_I("BDI-I");
const QString SCALE_BDI_IA("BDI-IA");
const QString SCALE_BDI_II("BDI-II");

const QStringList BDI_I_QUESTION_TOPICS{
    // from Beck 1988, https://doi.org/10.1016/0272-7358(88)90050-5
    "",  // index zero
    "mood",  // a
    "pessimism",  // b
    "sense of failure",  // c
    "lack of satisfaction",  // d
    "guilt feelings",  // e
    "sense of punishment",  // f
    "self-dislike",  // g
    "self-accusation",  // h
    "suicidal wishes",  // i
    "crying",  // j
    "irritability",  // k
    "social withdrawal",  // l
    "indecisiveness",  // m
    "distortion of body image",  // n
    "work inhibition",  // o
    "sleep disturbance",  // p
    "fatigability",  // q
    "loss of appetite",  // r
    "weight loss",  // s
    "somatic preoccupation",  // t
    "loss of libido",  // u
};
const QStringList BDI_IA_QUESTION_TOPICS = {
    // from [Beck1996b]
    "",  // index zero
    "sadness",  // 1
    "pessimism",
    "sense of failure",
    "self-dissatisfaction",
    "guilt",  // 5
    "punishment",
    "self-dislike",
    "self-accusations",
    "suicidal ideas",
    "crying",  // 10
    "irritability",
    "social withdrawal",
    "indecisiveness",
    "body image change",
    "work difficulty",  // 15
    "insomnia",
    "fatigability",
    "loss of appetite",
    "weight loss",
    "somatic preoccupation",  // 20
    "loss of libido",
};
const QStringList BDI_II_QUESTION_TOPICS = {
    // from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5889520/;
    // also https://www.ncbi.nlm.nih.gov/pubmed/10100838;
    // also [Beck1996b]
    // matches BDI-II paper version
    "",  // index zero
    "sadness",  // 1
    "pessimism",
    "past failure",
    "loss of pleasure",
    "guilty feelings",  // 5
    "punishment feelings",
    "self-dislike",
    "self-criticalness",
    "suicidal thoughts or wishes",
    "crying",  // 10
    "agitation",
    "loss of interest",
    "indecisiveness",
    "worthlessness",
    "loss of energy",  // 15
    "changes in sleeping pattern",  // decrease or increase
    "irritability",
    "changes in appetite",  // decrease or increase
    "concentration difficulty",
    "tiredness or fatigue",  // 20
    "loss of interest in sex",
};

void initializeBdi(TaskFactory& factory)
{
    static TaskRegistrar<Bdi> registered(factory);
}

Bdi::Bdi(CamcopsApp& app, DatabaseManager& db, const int load_pk) :
    Task(app, db, BDI_TABLENAME, false, false, false),  // ... anon, clin, resp
    m_grid_i(nullptr),
    m_grid_ia(nullptr),
    m_grid_ii(nullptr)
{
    addField(FN_BDI_SCALE, QMetaType::fromType<QString>());
    addFields(
        strseq(QPREFIX, FIRST_Q, N_QUESTIONS), QMetaType::fromType<int>()
    );

    load(load_pk);  // MUST ALWAYS CALL from derived Task constructor.
}

// ============================================================================
// Class info
// ============================================================================

QString Bdi::shortname() const
{
    return "BDI";
}

QString Bdi::longname() const
{
    return tr("Beck Depression Inventory");
}

QString Bdi::description() const
{
    return tr("21-item self-report scale (for BDI, BDI-1A, BDI-II).");
}

// ============================================================================
// Instance info
// ============================================================================

bool Bdi::isComplete() const
{
    return noneNull(values(strseq(QPREFIX, FIRST_Q, N_QUESTIONS)));
}

bool Bdi::isBdiII() const
{
    return valueString(FN_BDI_SCALE) == SCALE_BDI_II;
}

QStringList Bdi::summary() const
{
    using mathfunc::describeAsRanges;
    using stringfunc::bold;
    using stringfunc::strnumlist;

    const QString scale = valueString(FN_BDI_SCALE);
    const QVariant suicide_value = value(SUICIDALITY_FN);

    // Suicidal thoughts:
    QString suicide_description;
    if (suicide_value.isNull()) {
        suicide_description = "? (not completed)";
    } else {
        const int suicidality_score = suicide_value.toInt();
        suicide_description = QString::number(suicidality_score);
    }

    // Custom somatic score for Khandaker Insight study:
    QString somatic_text;
    if (isBdiII()) {
        const QStringList somatic_fieldnames
            = strnumlist(QPREFIX, CUSTOM_SOMATIC_KHANDAKER_BDI_II_QNUMS);
        const QVector<QVariant> somatic_values = values(somatic_fieldnames);
        bool somatic_missing = false;
        int somatic_score = 0;
        for (const QVariant& v : somatic_values) {
            if (v.isNull()) {
                somatic_missing = true;
                break;
            }
            somatic_score += v.toInt();
        }
        somatic_text
            = somatic_missing ? "incomplete" : QString::number(somatic_score);
    } else {
        somatic_text = "N/A";  // not the BDI-II
    }

    QString suicidality_topic;
    if (scale == SCALE_BDI_I) {
        suicidality_topic = BDI_I_QUESTION_TOPICS.at(SUICIDALITY_QNUM);
    } else if (scale == SCALE_BDI_IA) {
        suicidality_topic = BDI_IA_QUESTION_TOPICS.at(SUICIDALITY_QNUM);
    } else if (scale == SCALE_BDI_II) {
        suicidality_topic = BDI_II_QUESTION_TOPICS.at(SUICIDALITY_QNUM);
    } else {
        suicidality_topic = "suicidality";
    }

    // Summary:
    return QStringList{
        QString("Scale: %1.").arg(bold(valueString(FN_BDI_SCALE))),
        totalScorePhrase(totalScore(), MAX_QUESTION_SCORE),
        // Q9 is suicidal ideation in all versions of the BDI (I, IA, II).
        QString("Q%1 (%2): %3.")
            .arg(
                QString::number(SUICIDALITY_QNUM),
                suicidality_topic,
                bold(suicide_description)
            ),
        QString("Custom somatic score for Insight study "
                "(sum of scores for questions %1 for BDI-II only): %2.")
            .arg(
                describeAsRanges(CUSTOM_SOMATIC_KHANDAKER_BDI_II_QNUMS),
                bold(somatic_text)
            ),
    };
}

QStringList Bdi::detail() const
{
    return summary() + completenessInfo();
}

OpenableWidget* Bdi::editor(const bool read_only)
{
    const NameValueOptions options{
        {"0", 0},
        {"1", 1},
        {"2", 2},
        {"3", 3},
    };
    const NameValueOptions scale_options{
        {"BDI (1961; BDI-I)", SCALE_BDI_I},
        {"BDI-IA (1978)", SCALE_BDI_IA},
        {"BDI-II (1996)", SCALE_BDI_II},
    };
    QVector<QuestionWithOneField> fields_i;
    QVector<QuestionWithOneField> fields_ia;
    QVector<QuestionWithOneField> fields_ii;
    const QString& question_prefix = TextConst::question();
    for (int n = FIRST_Q; n <= N_QUESTIONS; ++n) {
        const QString qstrnum = QString::number(n);
        const QString fieldname = QPREFIX + qstrnum;
        const QString& topic_i = BDI_I_QUESTION_TOPICS.at(n);
        const QString& topic_ia = BDI_IA_QUESTION_TOPICS.at(n);
        const QString& topic_ii = BDI_II_QUESTION_TOPICS.at(n);
        const QString working_prefix = question_prefix + " " + qstrnum + " (";
        const QString question_i = working_prefix + topic_i + ")";
        const QString question_ia = working_prefix + topic_ia + ")";
        const QString question_ii = working_prefix + topic_ii + ")";
        fields_i.append(QuestionWithOneField(fieldRef(fieldname), question_i));
        fields_ia.append(QuestionWithOneField(fieldRef(fieldname), question_ia)
        );
        fields_ii.append(QuestionWithOneField(fieldRef(fieldname), question_ii)
        );
    }

    m_grid_i = new QuMcqGrid(fields_i, options);
    m_grid_ia = new QuMcqGrid(fields_ia, options);
    m_grid_ii = new QuMcqGrid(fields_ii, options);
    m_grid_i->addTag(SCALE_BDI_I);
    m_grid_ia->addTag(SCALE_BDI_IA);
    m_grid_ii->addTag(SCALE_BDI_II);

    if (valueIsNullOrEmpty(FN_BDI_SCALE)) {
        // first edit; set default
        setValue(FN_BDI_SCALE, SCALE_BDI_II);
    }

    // Set initial visibility:
    scaleChanged();

    // Callback
    FieldRefPtr fr_scale = fieldRef(FN_BDI_SCALE);
    connect(
        fr_scale.data(), &FieldRef::valueChanged, this, &Bdi::scaleChanged
    );

    QuPagePtr page(
        (new QuPage({
             (new QuText(appstring(appstrings::DATA_COLLECTION_ONLY)))
                 ->setBold(),
             new QuText(appstring(appstrings::BDI_WHICH_SCALE)),
             (new QuMcq(fr_scale, scale_options))
                 ->setHorizontal(true)
                 ->setAsTextButton(true),
             new QuText(TextConst::enterTheAnswers()),
             // Add all three grids; we'll swap between them.
             m_grid_i,
             m_grid_ia,
             m_grid_ii,
         }))
            ->setTitle(shortname())
    );

    auto questionnaire = new Questionnaire(m_app, {page});
    questionnaire->setType(QuPage::PageType::Patient);
    questionnaire->setReadOnly(read_only);
    return questionnaire;
}

// ============================================================================
// Task-specific calculations
// ============================================================================

int Bdi::totalScore() const
{
    return sumInt(values(strseq(QPREFIX, FIRST_Q, N_QUESTIONS)));
}

// ============================================================================
// Signal handlers
// ============================================================================

void Bdi::scaleChanged()
{
    if (!m_grid_i || !m_grid_ia || !m_grid_ii) {
        return;
    }
    const QString current_scale = valueString(FN_BDI_SCALE);
    // Initial visibility
    m_grid_i->setVisible(current_scale == SCALE_BDI_I);
    m_grid_ia->setVisible(current_scale == SCALE_BDI_IA);
    m_grid_ii->setVisible(current_scale == SCALE_BDI_II);
}
